osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1297 @@
+
+#include "osquery/config/parsers/ipmi.h"
+
+namespace osquery {
+
+// Callbacks for OpenIPMI
+extern "C" {
+void getLANsCB(ipmi_domain_t* domain, ipmi_mc_t* mc, void* data);
+void readThresholdSensorCB(ipmi_sensor_t* sensor,
+                           int err,
+                           enum ipmi_value_present_e value_present,
+                           unsigned int raw_value,
+                           double val,
+                           ipmi_states_t* states,
+                           void* data);
+void getFRUCB(ipmi_entity_t* entity, void* data);
+void iterateMCsCB(ipmi_domain_t* domain, ipmi_mc_t* mc, void* data);
+void IPMIFullyUpCB(ipmi_domain_t* domain, void* data);
+void ipmiLoggerCB(os_handler_t* handler,
+                  const char* format,
+                  enum ipmi_log_type_e logType,
+                  va_list ap);
+}
+
+/// Unique pointer deleter for os_handler_t*
+const auto kFreeOSHandle = [](os_handler_t* h) {
+  if (h != nullptr) {
+    ipmi_posix_free_os_handler(h);
+  }
+};
+
+/// Unique pointer deleter for ipmi_lanparm_t*
+const auto kFreeLANParm = [](ipmi_lanparm_t* lp) {
+  int rv = ipmi_lanparm_destroy(lp, NULL, NULL);
+  if (rv != 0) {
+    LOG(WARNING) << "Did not successfully destroy ipmi_lanparm_t*: "
+                 << strerror(rv);
+  }
+};
+
+/**
+ * @brief retrieves name of MC
+ *
+ * @param mc pointer to OPENIPMI ipmi_mc_t
+ * @param defaultName fallback name if the MC name is unable to be retrieved
+ *
+ * @return MC name as std::string
+ */
+static inline std::string getMCName(
+    ipmi_mc_t* mc, const std::string& defaultName = "missing") {
+  char name[IPMI_MC_NAME_LEN];
+  int len = ipmi_mc_get_name(mc, name, IPMI_MC_NAME_LEN);
+  if (len < 1) {
+    return defaultName;
+  }
+
+  return name;
+}
+
+/**
+ * @brief generates key to uniquely identify an MC instance
+ *
+ * @param mc pointer to OPENIPMI ipmi_mc_t
+ *
+ * @return unique key for MC as std::string
+ */
+static inline std::string getMCKey(ipmi_mc_t* mc) {
+  return getMCName(mc) + "-" + std::to_string(ipmi_mc_device_id(mc));
+}
+
+class IPMIClient;
+
+/**
+ * @brief parmData encapsulates things that are required for retrieving LANPARM
+ *
+ * @param client pointer to IPMIClient instance
+ * @param mcKey unique key to identify the MC
+ * @parm colName the osquery table column name of the LANPARM
+ *
+ * @parm parm OpenIPMI identifier for the LANPARM
+ *
+ */
+struct parmData {
+  IPMIClient* client;
+  std::string mcKey;
+  std::string colName;
+  unsigned int parm;
+
+  parmData() : client(nullptr), mcKey(""), colName(""), parm(999) {}
+};
+
+/**
+ * @brief Client for interacting with OpenIPMI
+ *
+ * IPMIClient is a singleton object accessible only by the IPMIClient::get
+ * static method.  IPMIClient is lazily loaded on the first query to any IPMI
+ * table.  The initialization process waits for OpenIPMI to be in the "fully up"
+ * state so it can a few minutes.
+ */
+class IPMIClient : public InternalRunnable {
```

Comment:
Yeah I was thinking that originally, but kept it there b/c I didn't think it would be used anywhere else.  But that was before the move to put the implementation in its own `ipmi` directory under `osquery/tables`.  It makes a lot more sense to do that now.  Good suggestion! :)